### PR TITLE
Async interceptor payload + S3 Checksums

### PR DESCRIPTION
**Pull request by spfink**
Exposes the ability to modify the AsyncRequestBody and AsyncResponseTransformer.

Currently this is done by adding two new interceptors.

For AsyncRequestBody:

```java
default AsyncRequestBody modifyAsyncRequestBody(Context.ModifyRequest context,
                                                ExecutionAttributes executionAttributes) {
    return context.requestBody();
}
```

For AsyncResponseTransformer:

```java
default AsyncResponseTransformer modifyAsyncResponseTransformer(Context.ModifyRequest context,
                                                                ExecutionAttributes executionAttributes) {
    return context.responseTransformer();
}
```

Possible other options include making modifyRequest return back a container object instead of an SdkRequest:

```java
default SdkRequestContext modifyRequest(Context.ModifyRequest context, ExecutionAttributes executionAttributes) {
    return context.request();
}
```

Where SdkRequestContext contains SdkRequest, AsyncRequestBody and AsyncResponseTransformer.

Or we could introduce an AsyncExecutionInterceptor interface that again either has two methods as in the PR or one method with a container object for async.

In general, trying to avoid potential confusion of when these interceptors are valid and when the values will be populated. If we use the approach in the PR, users of the Sync client should not use the new interceptors. If we go with the container approach for modifyRequest, then the asyncRequestBody and asyncResponseTransformer will be null.

### CODE REVIEW

**Comment by millems** [comment]
The sync payload is currently on the HTTP request, and interceptors modify the HTTP request payload by wrapping the contentStreamProvider in modifyHttpRequest.

Why can't we do the same for async?

**Comment by zoewangg** [comment]
@shorea +1 on your first option. Like the idea of splitting sync and async modifyHttpContents methods.

Would it make sense to further split the `Context`?
`Context.ModifyHttpContent` and `Context.ModifyAsyncHttpContent`

For the second option, I feel it could be a bit confusing because the it can get the inputStream from both `toModify` and `context.httpContent()`, even though `context.httpContent()` would be null at that time, but the person who implements this method  might not know.

```java
InputStream modifyHttpContent(InputStream toModify, Context.ModifyHttpContent context, ExecutionAttributes executionAttributes);
```

**Comment by shorea** [comment]
So I was proposing not to have the content in the context at all until after that modifyHttpContent stage, so it's only in one location (the parameter in modifyHttpContent and the next context object in future callbacks).

**Comment by millems** [comment]
As long as sync and async are consistent, I don't mind too much. 

1. If we're going to have sync payloads on the HTTP request, we should have async payloads there as well. 
2. If we use AsyncRequestBody in the interceptor for async, we should use RequestBody for async.

**Comment by spfink** [comment]
Updated PR to include separate content interceptors.

@millems, not sure I follow your latest comment. Are you saying we should put the response transformer and request body on the SdkHttpFullRequest/Response objects?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java [hunk 1]
```diff
@@ -204,6 +205,7 @@ protected SdkClientConfiguration finalizeChildConfiguration(SdkClientConfigurati
     private SdkClientConfiguration finalizeSyncConfiguration(SdkClientConfiguration config) {
         return config.toBuilder()
                      .option(SdkClientOption.SYNC_HTTP_CLIENT, resolveSyncHttpClient(config))
+                     .option(SdkClientOption.CLIENT_TYPE, "Sync")
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java]: Enum?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java [hunk 2]
```diff
@@ -65,19 +69,35 @@ private Context() {}
...
+     */
+    @ThreadSafe
+    @SdkPublicApi
+    public interface ModifyHttpContent extends ModifyHttpRequest {
         /**
          * The {@link SdkHttpFullRequest} that was created as a result of marshalling the {@link #request()}. This is the HTTP
          * request that will be sent to the downstream service.
          */
-        SdkHttpFullRequest httpRequest();
+        SdkHttpRequest httpRequest();
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: Shouldn't this be in AfterMarshalling?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java [hunk 3]
```diff
@@ -86,6 +106,9 @@ private Context() {}
     @ThreadSafe
     @SdkPublicApi
     public interface BeforeTransmission extends ModifyHttpRequest {
+        RequestBody requestBody();
+
+        AsyncRequestBody asyncRequestBody();
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: Shouldn't this be in AfterMarshalling?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java [hunk 4]
```diff
@@ -105,15 +144,16 @@ private Context() {}
      */
     @ThreadSafe
     @SdkPublicApi
-    public interface ModifyHttpResponse extends AfterTransmission {
+    public interface ModifyHttpResponseContent extends ModifyAsyncHttpResponseContent {
+        Publisher responsePublisher();
     }
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: No raw generics, please.
Shouldn't this be in AfterTransmision?
Where is the sync payload?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionInterceptor.java [hunk 5]
```diff
@@ -15,17 +15,22 @@
...
+import software.amazon.awssdk.core.async.AsyncRequestBody;
 import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.core.sync.RequestBody;
 import software.amazon.awssdk.http.SdkHttpFullRequest;
 import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.http.SdkHttpRequest;
 
+// Disable CS to avoid "Unused Import" error. If we use the FQCN in the Javadoc, we'll run into line length issues instead.
+// CHECKSTYLE:OFF
+// CHECKSTYLE:ON
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionInterceptor.java]: Remove comments?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionInterceptor.java [hunk 6]
```diff
@@ -219,12 +238,49 @@ default void afterTransmission(Context.AfterTransmission context, ExecutionAttri
...
+     * the error code returned by the service is retriable, this will be invoked for each response returned by the service.
+     *
+     * @param context The current state of the execution, including the SDK and HTTP requests and the current HTTP response.
+     * @param executionAttributes A mutable set of attributes scoped to one specific request/response cycle that can be used to
+     * give data to future lifecycle methods.
+     * @return The potentially-modified HTTP response that should be given to the unmarshaller. Must not be null.
+     */
+    default Publisher<ByteBuffer> modifyAsyncHttpResponseContent(Publisher<ByteBuffer> toModify,
+                                                                 Context.ModifyAsyncHttpResponseContent context,
+                                                                 ExecutionAttributes executionAttributes) {
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionInterceptor.java]: The values to be modified aren't usually passed separately. Any reason we're changing that just for these?

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/GetBucketPolicyInterceptor.java [hunk 7]
```diff
@@ -34,18 +34,18 @@
...
-    public SdkHttpFullResponse modifyHttpResponse(Context.ModifyHttpResponse context, ExecutionAttributes executionAttributes) {
+    public RequestBody modifyHttpContent(RequestBody requestBody,
+                                          Context.ModifyHttpContent context,
+                                          ExecutionAttributes executionAttributes) {
         if (context.request() instanceof GetBucketPolicyRequest) {
-            String policy = context.httpResponse().content()
-                                   .map(c -> invokeSafely(() -> IoUtils.toUtf8String(c)))
-                                   .orElse(null);
+
+            String policy = requestBody.contentStreamProvider() == null ? null : invokeSafely(() ->IoUtils.toUtf8String(requestBody.contentStreamProvider().newStream()));
```
Comment (millems) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/GetBucketPolicyInterceptor.java]: run checkstyle

File: http-client-spi/src/main/java/software/amazon/awssdk/http/ExecuteRequest.java [hunk 8]
```diff
@@ -49,20 +56,29 @@ public static Builder builder() {
          * @param request The request.
          * @return This builder for method chaining.
          */
-        Builder request(SdkHttpFullRequest request);
+        Builder request(SdkHttpRequest request);
+
+        Builder contentStreamProvider(Optional<ContentStreamProvider> contentStreamProvider);
```
Comment (millems) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/ExecuteRequest.java]: Don't accept an optional parameter.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/InvokeableHttpRequest.java [hunk 9]
```diff
@@ -26,5 +26,5 @@
 @SdkPublicApi
 public interface InvokeableHttpRequest extends Callable<SdkHttpFullResponse>, Abortable {
     @Override
-    SdkHttpFullResponse call() throws IOException;
+    SdkHttpFullResponse call() throws IOException; // ExecuteResponse - sdkResponse + inputStream
```
Comment (millems) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/InvokeableHttpRequest.java]: TODO

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java (hunk 1 above)
Comment (spfink) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java]: Will add.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java (hunk 2 above)
Comment (spfink) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: See Andrew's comment on composite interceptors.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java (hunk 3 above)
Comment (spfink) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: Not if we have them as separate arguments in interceptors.

**Comment by millems** [comment]
@zoewangg @shorea The problem with removing http request from the AfterMarshalling context and http response from the AfterTransmission context, is that it makes afterMarshalling and afterTransmission much less useful. Those methods exist to see those values, so removing them means there's not really any reason to use them (or any way to see the values before they're modified by other interceptors).

If we moved them to the first parameter of afterMarshalling (eg.) it would fix the problem, but three parameters for every method sounds like a lot...

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java [hunk 10]
```diff
@@ -98,6 +124,10 @@ private Context() {}
          * The HTTP response returned by the service with which the SDK is communicating.
          */
         SdkHttpFullResponse httpResponse();
+
+        Publisher<ByteBuffer> responsePublisher();
+
+        InputStream responseBody();
```
[the remark this answers is not in the document]
Comment (spfink) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: We can make these both be Optional but then we'd likely want to make the interceptors that use them return optional as well.

[the remark this answers is not in the document]
Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/Context.java]: That makes sense to me! Good call.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/ClientType.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core;
+
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+@SdkPublicApi
+public enum ClientType {
```
[the remark this answers is not in the document]
Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/ClientType.java]: Ping - Javadoc

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/ClientExecutionParams.java [hunk 12]
```diff
@@ -80,6 +82,15 @@ public InputT getInput() {
         return this;
     }
 
+    public RequestBody getRequestBody() {
+        return requestBody;
+    }
+
+    public ClientExecutionParams<InputT, OutputT> withRequestBody(RequestBody requestBody) {
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/ClientExecutionParams.java]: This is a protected API with with*/get*. Can we fix that really quick?

File: http-client-spi/src/main/java/software/amazon/awssdk/http/InvokeableHttpRequest.java (hunk 9 above)
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/InvokeableHttpRequest.java]: Ping TODO

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullResponse.java [hunk 13]
```diff
@@ -166,5 +167,8 @@ default Builder putHeader(String headerName, String headerValue) {
          * This is usually done by closing the service connection.</p>
          */
         Builder content(AbortableInputStream content);
+
```
Comment (millems) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullResponse.java]: Are all the methods on this builder still needed?

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java [hunk 14]
```diff
@@ -122,9 +122,9 @@ private HttpRequestBase wrapEntity(SdkHttpFullRequest request,
          * preparation for the retry. Eventually, these wrappers would
          * return incorrect validation result.
          */
-        if (request.contentStreamProvider().isPresent()) {
+        if (request.contentStreamProvider() != null && request.contentStreamProvider().isPresent()) {
```
[the remark this answers is not in the document]
Comment (millems) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java]: Ping - why is contentStreamProvider null?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientSpiVerificationTest.java [hunk 15]
```diff
@@ -136,17 +136,17 @@ private SdkHttpFullRequest createRequest(URI endpoint,
...
-                                 .method(method)
-                                 .encodedPath(resourcePath)
-                                 .applyMutation(b -> params.forEach(b::putRawQueryParameter))
-                                 .applyMutation(b -> {
-                                     b.putHeader("Host", endpoint.getHost());
-                                     if (contentLength != null) {
-                                         b.putHeader("Content-Length", contentLength);
-                                     }
-                                 }).build();
+        return (SdkHttpFullRequest) SdkHttpFullRequest.builder()
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientSpiVerificationTest.java]: Why is this (and other similar casts) needed?

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/GetObjectIntegrationTest.java [hunk 16]
```diff
@@ -25,6 +25,7 @@
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.nio.file.Path;
+import org.apache.log4j.BasicConfigurator;
```
Comment (millems) [inline services/s3/src/it/java/software/amazon/awssdk/services/s3/GetObjectIntegrationTest.java]: Unused import

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/ClientExecutionParams.java (hunk 12 above)
Comment (spfink) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/ClientExecutionParams.java]: Will update right before I merge.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullResponse.java (hunk 13 above)
Comment (spfink) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullResponse.java]: Yes as discussed.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java [hunk 17]
```diff
@@ -136,7 +136,7 @@ private static JsonMarshallerRegistry createMarshallerRegistry() {
     }
 
     private SdkHttpFullRequest.Builder fillBasicRequestParams(OperationInfo operationInfo) {
-        return ProtocolUtils.createSdkHttpRequest(operationInfo, endpoint)
+        return (SdkHttpFullRequest.Builder) ProtocolUtils.createSdkHttpRequest(operationInfo, endpoint)
```
Comment (millems) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java]: Needed cast?

File: http-client-spi/src/main/java/software/amazon/awssdk/http/InvokeableHttpRequest.java [hunk 18]
```diff
@@ -26,5 +26,5 @@
 @SdkPublicApi
 public interface InvokeableHttpRequest extends Callable<SdkHttpFullResponse>, Abortable {
     @Override
-    SdkHttpFullResponse call() throws IOException;
+    SdkHttpFullResponse call() throws IOException; // TODO: Change to container instad of SdkHttpFullResponse
```
Comment (millems) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/InvokeableHttpRequest.java]: I meant, can we actually do this?

File: http-client-spi/src/test/java/software/amazon/awssdk/http/SdkHttpRequestResponseTest.java [hunk 19]
```diff
@@ -416,7 +416,7 @@ private SdkHttpFullRequest validRequest() {
     }
 
     private SdkHttpFullRequest.Builder validRequestBuilder() {
-        return SdkHttpFullRequest.builder().protocol("http").host("localhost").method(SdkHttpMethod.GET);
+        return (SdkHttpFullRequest.Builder) SdkHttpFullRequest.builder().protocol("http").host("localhost").method(SdkHttpMethod.GET);
```
Comment (millems) [inline http-client-spi/src/test/java/software/amazon/awssdk/http/SdkHttpRequestResponseTest.java]: Cast still needed?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientWireMockTest.java [hunk 20]
```diff
@@ -342,17 +342,17 @@ private SdkHttpFullRequest createRequest(URI uri,
...
-                                 .method(method)
-                                 .encodedPath(resourcePath)
-                                 .applyMutation(b -> params.forEach(b::putRawQueryParameter))
-                                 .applyMutation(b -> {
-                                     b.putHeader("Host", uri.getHost());
-                                     if (contentLength != null) {
-                                         b.putHeader("Content-Length", contentLength);
-                                     }
-                                 }).build();
+        return (SdkHttpFullRequest) SdkHttpFullRequest.builder()
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientWireMockTest.java]: Cast still needed?

**Pull request merged**